### PR TITLE
some fixes for 32bit builds

### PR DESCRIPTION
**Pull request by mcirsta**
This fixes part of the problems for 32bit builds ( issue 68 )

### CODE REVIEW

**Comment by sjackman** [comment]
My preference would be to consistently use `size_t` everywhere, and hack around that `1<<40` issue.

**Comment by sjackman** [comment]
Replacing `1<<40` with `1ULL << 40` might do the trick.

**Comment by benvvalk** [comment]
Yes, I like using size_t too.  Alternatively, I could just take out the option of using a "T" suffix for terabyte-size Bloom filter sizes.  That might fix it too.

**Comment by mcirsta** [comment]
The problem with size_t is that there's no guarantee of the size of size_t whereas uint64_t guarantees a 8 byte size no matter what.
 What I do consider bad practive are operations like in Bloom/Bloom.h where an unsigned long is compared to size_t. That only happened to work on 64bit but it will probably work nowhere else and I do find uint64_t provides consistency here.

 Is there a reason to avoid uint64_t ? It's definde in  C++11 as far as I know.

**Comment by sjackman** [comment]
`size_t` is sufficiently wide to address the maximum amount of memory on the host machine. On a 32-bit machine `size_t` will be more efficient than `uint64_t`. A `uint64_t` is unnecessarily wide, since the host machine is only able to address a maximum of 4 GB of RAM.

**Comment by mcirsta** [comment]
I fully agree with this but it's all down to what values you inted to keep in that variable. For memory locations of size of a string and other things like that size_t is more than enough but if for instance you want to keep track of the number of sand particles on a beach there's not guarantee a 32 bit size_t will be large enough ( well there's not guarantee a 64 bit one will be but this is just an example ).
 
 I've no idea what values you have in this software and how large they can be ... if they're not larger than max(4 byte) and will not be so in the future then indeed size_t is good enough.

**Comment by sjackman** [comment]
We do in fact use `uint64_t` when we're counting, for example, the number of *k*-mer observed, but use `size_t` when counting the number of distinct *k*-mer, which can't exceed the amount of addressable memory.

**Comment by sjackman** [comment]
In any case, your PR certainly helps highlight areas of code that need to reviewed to see if they've used the right type (`uint64_t` or `size_t`). We may not merge this PR, but we'll commit something to fix your compiler error.

File: Bloom/Bloom.h [hunk 1]
```diff
@@ -51,7 +52,7 @@ namespace Bloom {
 	/** file format version number */
 	static const unsigned BLOOM_VERSION = 2;
 	/** I/O buffer size when reading/writing bloom filter files */
-	static const unsigned long IO_BUFFER_SIZE = 32*1024;
+	static const uint64_t IO_BUFFER_SIZE = 32*1024;
```
Comment (sjackman) [inline Bloom/Bloom.h]: Use `size_t`.

File: Bloom/Bloom.h [hunk 2]
```diff
@@ -165,7 +166,7 @@ namespace Bloom {
 		// bloom filter bits
 
 		size_t bits = endBitPos - startBitPos + 1;
-		size_t bytes = (bits + 7) / 8;
+		uint64_t bytes = (bits + 7) / 8;
```
Comment (sjackman) [inline Bloom/Bloom.h]: Use `size_t`.

File: Common/StringUtil.h [hunk 3]
```diff
@@ -106,7 +107,7 @@ bool isReadNamePair(const std::string& name1, const std::string& name2)
 	return false;
 }
 
-static inline size_t SIToBytes(std::istringstream& iss)
+static inline uint64_t SIToBytes(std::istringstream& iss)
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

File: Common/StringUtil.h [hunk 4]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
 
 	switch(tolower(units[0])) {
 		case 'k':
-			size *= (size_t)1<<10; break;
+			size *= (uint64_t)1<<10; break;
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

File: Common/StringUtil.h [hunk 5]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
...
 		case 'm':
-			size *= (size_t)1<<20; break;
+			size *= (uint64_t)1<<20; break;
 		case 'g':
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

File: Common/StringUtil.h [hunk 6]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
 
 	switch(tolower(units[0])) {
 		case 'k':
-			size *= (size_t)1<<10; break;
+			size *= (uint64_t)1<<10; break;
 		case 'm':
-			size *= (size_t)1<<20; break;
+			size *= (uint64_t)1<<20; break;
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

File: Common/StringUtil.h [hunk 7]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
...
 		case 'g':
-			size *= (size_t)1<<30; break;
+			size *= (uint64_t)1<<30; break;
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

File: Common/StringUtil.h [hunk 8]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
...
 		case 't':
-			size *= (size_t)1<<40; break;
+			size *= (uint64_t)1<<40; break;
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t` and change `1<<40` to `1ULL << 40`.

File: Common/StringUtil.h [hunk 9]
```diff
@@ -133,19 +134,19 @@ static inline size_t SIToBytes(std::istringstream& iss)
...
 		default:
 			iss.setstate(std::ios::failbit);
 			return 0;
 	}
 
-	return (size_t)ceil(size);
+	return (uint64_t)ceil(size);
```
Comment (sjackman) [inline Common/StringUtil.h]: Use `uint64_t`.

**Comment by benvvalk** [comment]
Thanks for your help, @mcirsta and @sjackman.  The Bloom.h code is mine, and I understand now why I need to be more careful about my use of unsigned long / uint64_t / size_t.   I will fix the code.

The purpose of the variable is to store the byte size of Bloom filters in memory, so size_t  is the most appropriate choice.

**Comment by mcirsta** [comment]
Thanks, the truth is I'm fine with size_t too if it's large enough for the purpose, and ultimately you know the code better than I do , the changes proposed were just a way to fix some compile errors by someone who has no idea what the code does.
 
 This has been a good learning experience for me too as some things are not really obvious as far as C/C++ types go.

 The final decision on what to merge is yours and I've no problem with that.

**Comment by sjackman** [comment]
@mcirsta Has this issue been resolved, that is, does ABySS compile cleanly on 32 bit architectures?

**Pull request closed without merging**